packages/kit/src/runtime/client/router.js

```diff
@@ -55,12 +55,26 @@ export class Router {
 		// make it possible to reset focus
 		document.body.setAttribute('tabindex', '-1');
 
+		history.scrollRestoration = 'manual';
+
 		// keeping track of the history index in order to prevent popstate navigation events if needed
-		this.current_history_index = history.state?.['sveltekit:index'] ?? 0;
+		/** @type {number} */
+		this.current_history_index = Number(
+			sessionStorage.getItem('sveltekit:index') ?? history.state?.['sveltekit:index'] ?? 0
+		);
 
-		if (this.current_history_index === 0) {
+		if (!history.state || history.state['sveltekit:index'] == undefined) {
 			// create initial history entry, so we can return here
-			history.replaceState({ ...history.state, 'sveltekit:index': 0 }, '', location.href);
+			history.replaceState(
+				{ ...(history.state || {}), 'sveltekit:index': this.current_history_index },
```

Comment:
it's fine to spread `undefined`

```suggestion
				{ ...history.state, 'sveltekit:index': this.current_history_index },
```